user_interface/vscode-nupad/src/extension.ts

```diff
@@ -0,0 +1,140 @@
+// The module 'vscode' contains the VS Code extensibility API
+// Import the module and reference it with the alias vscode in your code below
+import * as vscode from 'vscode';
+import * as WebSocket from 'ws';
+
+const INIT_REQUEST = 1;
+const INIT_RESPONSE = 2;
+const UI_INPUT = 3;
+const REMOTE_CHANGE = 4;
+const INSERT_OPERATION = 1;
+const DELETE_OPERATION = 2;
+const CONNECT_URL = 'ws://127.0.0.1:9002';
+
+class NUpadClient {
+
+    title: string;
+    peerName: string;
+    socket: WebSocket;
+    remoteChange: boolean;
+
+    constructor() {
+        this.title = "demo-doc";
+        this.peerName = "";
+        this.socket = this.initSocket();
+        this.remoteChange = false;
+    }
+
+    initSocket = () => {
+        let socket = new WebSocket(CONNECT_URL);
+        socket.onopen = () => {
+            this.sendMessage({
+                message_type: INIT_REQUEST,
+                document_name: this.title
+            });
+        };
+        socket.onclose = () => {
+            console.log("closed");
+        };
+        socket.onmessage = this.handleRemoteChange;
+        return socket;
+    };
+
+    sendMessage = (message: any) => {
+        console.log("sending: ", message);
+        this.socket.send(JSON.stringify(message));
+    };
+
+    setContent = (content: string) => {
+        const editor = vscode.window.activeTextEditor;
+        if (editor) {
+            const document = editor.document;
+            const textRange = new vscode.Range(
+                document.positionAt(0),
+                document.positionAt(document.getText().length)
+            );
+
+            editor.edit(editBuilder => {
+                const firstLine = document.lineAt(0);
+                editBuilder.delete(textRange);
+                editBuilder.insert(firstLine.range.start, content);
+            });
+        } else {
+            console.log("cannot find text editor");
+        }
+    };
+
+    handleRemoteChange = (e: any) => {
+        this.remoteChange = true;
+        const msg = JSON.parse(e.data);
+        console.log("Received -> ", msg);
+        if (msg['messageType'] === INIT_RESPONSE) {
+            this.peerName = msg['peerName'];
+            this.setContent(msg['initialContent']);
+        } else if (msg['messageType'] === REMOTE_CHANGE) {
+            this.setContent(msg['content']);
+        } else {
+            console.error("unknown message received: ", msg);
+        }
+    };
+
+    handleDocumentChange = (e: any) => {
+        console.log("event", e);
+        if (this.remoteChange) {
```

Comment:
could there be a race condition here? Both the `handleRemoteChange` and `handleDocumentChange` events can happen simultaneously
the `socket.onmessage` also calls `handleRemoteChange` method, which sets `this.remoteChange=true`.

If just after that time, a local change is made, which calls the `handleDocumentChange` method, and this will see `this.remoteChange` to `true`, which will then return incorrectly without sending the appropriate message to nupad_app server

Reply:
resolved offline

Reply:
This cannot happen since all events will be processed inside a single event loop.
Hence it is recommended to not perform CPU-intensive tasks inside the event loop, those tasks should be offloaded to Promises.